### PR TITLE
JIT: import entire method for OSR, prune unneeded parts later

### LINKED ISSUE #83783: JIT: OSR is not conservative enough about exposing struct locals
```csharp
using System;
using System.Runtime.CompilerServices;

public class MissingExposure
{
    [MethodImpl(MethodImplOptions.NoInlining)]
    static void Bar()
    {
    }

    [MethodImpl(MethodImplOptions.NoInlining)]
    static void Foo(int n)
    {
        S s = new S { F = 1234 };
        ref int foo = ref s.F;

        for (int i = 0; i < n; i++)
        {
            Bar();
        }

        int abc = s.F * 3 + 4;
        foo = 25;
        int def = s.F * 3 + 4;
        Console.WriteLine("abc = {0} (expected {1}), def = {2} (expected {3})", abc, 1234 * 3 + 4, def, 25 * 3 + 4);
    }

    public static void Main()
    {
        Foo(50000);
    }

    public struct S
    {
        public int F;
    }
}
```

Output with OSR enabled: abc = 3706 (expected 3706), def = 3706 (expected 79)

cc @AndyAyersMS

### PR DESCRIPTION
**Pull request by AndyAyersMS**
For OSR compiles, always import from the original entry point in addtion to the OSR entry point. This gives the OSR compiler a chance to see all of the method and so properly compute address exposure, instead of relying on the Tier0
analysis.

Once address exposure has been determined, revoke special protection for the original entry and try and prune away blocks that are no longer needed.

Fixes #83783.

May also fix some of the cases where OSR perf is lagging (though don't expect this to fix them all).

### CODE REVIEW

**Comment by AndyAyersMS** [comment]
FYI @jakobbotsch 
cc @dotnet/jit-contrib 

There are several ways to approach this -- in this version the original entry is unreachable, but we force it to stay around; alternatively, we could make the OSR entry unreachable and force it to stay around, and then swap entry points later.

This one is a bit less disruptive; if it pans out I can also reconcile it with the code where in OSR mode we forcibly import the original entry if we think the method might tail call (and which is why I know this approach will "work", since we already do this extra importation some of the time). For tail call cases we keep the original entry protected until after morph, so a bit later than what we do here. The protection mechanisms are different, so the tail call version still works.

This will be difficult to assess via SPMI as many OSR contexts will fail to replay. But some of them do and the diffs generally look encouraging. 

It will also hurt OSR TP but again this may be hard to spot in SPMI. In actuality OSR compiles are pretty rare so I'm not that worried about the extra work.

A third option is to just compile the method normally and then once we're past morph say, rework the control flow to jump to the OSR entry from scratch. That would be more work because I'd have to revise the whole "OSR step block" scheme so that it could run later, in case the OSR entry happens to be in the middle of a bunch of try regions.

Much of what happens before morph is currently flow insensitive, so it doesn't matter that a big swath of blocks are unreachable. It might mess up early liveness, but if so we already have this problem in methods that might tail call.

This may fix some latent OSR perf issues, but I don't expect it to fix them all.

**Comment by AndyAyersMS** [comment]
Hmm, looks like a fairly persistent assert.

**Comment by AndyAyersMS** [comment]
Problem is that once you set `BBF_DONT_REMOVE` it is hard to safely un-set it. So I will just revise and use the existing artificial ref count solution. This will keep the extra IR and blocks around a bit longer but be simpler overall.

**Comment by AndyAyersMS** [comment]
Test failures seem unrelated. Passed jit-experimental which has various forms of OSR stress.

**Review by jakobbotsch — APPROVED** [review]
LGTM! I thought this would be a much more involved change

File: src/coreclr/jit/lclvars.cpp [hunk 1]
```diff
@@ -334,19 +334,6 @@ void Compiler::lvaInitTypeRef()
         {
             LclVarDsc* const varDsc = lvaGetDesc(lclNum);
             varDsc->lvIsOSRLocal    = true;
-
-            if (info.compPatchpointInfo->IsExposed(lclNum))
-            {
```
Comment (jakobbotsch) [inline src/coreclr/jit/lclvars.cpp:339]: Do we still need to communicate the exposure information in the patchpoint information?

**Comment by AndyAyersMS** [comment]
> LGTM! I thought this would be a much more involved change

Luckily (I guess) we already had this capability for some OSR methods, so now we just use it for all of them.

File: src/coreclr/jit/lclvars.cpp (hunk 1 above)
Comment (AndyAyersMS) [reply src/coreclr/jit/lclvars.cpp:339]: No. However, I'll leave this as is for now so older jits can still do somewhat reasonable things with newer SPMI data.

**Pull request merged**